rules/ast/ast.go

```diff
@@ -749,55 +756,185 @@ func labelsEqual(labels1, labels2 clientmodel.Metric) bool {
+		lhs := node.lhs.(VectorNode).Eval(timestamp)
 		rhs := node.rhs.(VectorNode).Eval(timestamp)
-		for _, rhsSample := range rhs {
-			value, keep := evalVectorBinop(node.opType, lhs, rhsSample.Value)
-			if keep {
-				rhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					rhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-				}
-				result = append(result, rhsSample)
+
+		return node.evalVectors(timestamp, lhs, rhs)
+	}
+
+	// vector-to-scalar operation
+	var lhs Vector
+	var rhs clientmodel.SampleValue
+	swap := false
+
+	if node.lhs.Type() == ScalarType && node.rhs.Type() == VectorType {
+		lhs = node.rhs.(VectorNode).Eval(timestamp)
+		rhs = node.lhs.(ScalarNode).Eval(timestamp)
+		swap = true
+	} else {
+		lhs = node.lhs.(VectorNode).Eval(timestamp)
+		rhs = node.rhs.(ScalarNode).Eval(timestamp)
+	}
+
+	result := make(Vector, 0, len(lhs))
+
+	for _, lhsSample := range lhs {
+		lv, rv := lhsSample.Value, rhs
+		// lhs always contains the vector. If the original position was different
+		// swap for calculating the value.
+		if swap {
+			lv, rv = rv, lv
+		}
+		value, keep := evalVectorBinop(node.opType, lv, rv)
+		if keep {
+			lhsSample.Value = value
+			if node.opType.shouldDropMetric() {
+				lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
 			}
+			result = append(result, lhsSample)
 		}
-		return result
-	} else if node.lhs.Type() == VectorType && node.rhs.Type() == ScalarType {
-		lhs := node.lhs.(VectorNode).Eval(timestamp)
-		rhs := node.rhs.(ScalarNode).Eval(timestamp)
-		for _, lhsSample := range lhs {
-			value, keep := evalVectorBinop(node.opType, lhsSample.Value, rhs)
-			if keep {
-				lhsSample.Value = value
-				if node.opType.shouldDropMetric() {
-					lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-				}
-				result = append(result, lhsSample)
+	}
+	return result
+}
+
+// evalVectorBinop evaluates the binary operation for the given vectors.
+func (node *VectorArithExpr) evalVectors(timestamp clientmodel.Timestamp, lhs, rhs Vector) Vector {
+	result := make(Vector, 0, len(rhs))
+
+	// the control flow below handles one-to-one or many-to-one matching.
+	// for one-to-many swap sidedness and account for the swap when calculating
+	// values.
+	if node.matchType == MatchOTM {
+		lhs, rhs = rhs, lhs
+	}
+
+	// all samples from the rhs hashed by the matching label/values
+	rm := make(map[clientmodel.Fingerprint]*Sample)
+	// contains all fingerprints of rhs samples that have been used in an operation
+	// with a lhs sample. It is used to detect illegal many-to-one matchings.
+	added := make(map[clientmodel.Fingerprint]bool)
+
+	// add all rhs samples to a map so we can easily find matches later
+	for _, rs := range rhs {
+		fp := node.fingerprint(rs.Metric)
+		// the rhs is guaranteed to be the 'one' side. Having multiple samples
+		// with the same fingerprint means that the matching is many-to-many
+		// which is not supported
+		if rm[fp] != nil {
+			panic("many to many matching not allowed")
```

Comment:
Is this an error that can happen from wrong data / user input? We should never panic on user errors... panics are only for responding to actual bugs in Prometheus code that are completely unexpected (assert-style).

The user won't see this error propagated to them in an interactive query, and a panic in a rule will kill the entire Prometheus server.

Currently we don't return errors from the AST `Eval()` nodes, but cope by just returning "special" results, like `NaN` for a `scalar()` of a vector with not exactly one time series. Can something similar be done here? If not, it might be time to introduce error returns from the AST nodes.